src/sdk/pynni/nni/compression/torch/pruning/amc/channel_pruning_env.py

```diff
@@ -0,0 +1,497 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+import os
+import time
+import math
+import copy
+import numpy as np
+import torch
+import torch.nn as nn
+
+from nni.compression.torch.compressor import PrunerModuleWrapper
+from .lib.utils import prGreen
+from .. import AMCWeightMasker
+
+# for pruning
+def acc_reward(net, acc, flops):
+    return acc * 0.01
+
+
+def acc_flops_reward(net, acc, flops):
+    error = (100 - acc) * 0.01
+    return -error * np.log(flops)
+
+
+class ChannelPruningEnv:
+    """
+    Env for channel pruning search
+    """
+    def __init__(self, pruner, val_func, val_loader, checkpoint, args):
+        self.pruner = pruner
+        self.model = pruner.bound_model
+        self.checkpoint = checkpoint
+        self.batch_size = val_loader.batch_size
+        self.preserve_ratio = args.preserve_ratio
+        self.channel_prune_masker = AMCWeightMasker(self.model, self.pruner, args.channel_round)
+
+        # options from args
+        self.args = args
+        self.lbound = args.lbound
+        self.rbound = args.rbound
+
+        self.n_calibration_batches = args.n_calibration_batches
+        self.n_points_per_layer = args.n_points_per_layer
+        self.channel_round = args.channel_round
+
+        # sanity check
+        assert self.preserve_ratio > self.lbound, 'Error! You can make achieve preserve_ratio smaller than lbound!'
+
+        # prepare data
+        self._val_loader = val_loader
+        self._validate = val_func
+
+        # build indexs
+        self._build_index()
+        self.n_prunable_layer = len(self.prunable_idx)
+
+        # extract information for preparing
+        self._extract_layer_information()
+
+        # build embedding (static part)
+        self._build_state_embedding()
+
+        # build reward
+        self.reset()  # restore weight
+        self.org_acc = self._validate(self._val_loader, self.model)
+        print('=> original acc: {:.3f}%'.format(self.org_acc))
+        self.org_model_size = sum(self.wsize_list)
+        print('=> original weight size: {:.4f} M param'.format(self.org_model_size * 1. / 1e6))
+        self.org_flops = sum(self.flops_list)
+        print('=> FLOPs:')
+        print([self.layer_info_dict[idx]['flops']/1e6 for idx in sorted(self.layer_info_dict.keys())])
+        print('=> original FLOPs: {:.4f} M'.format(self.org_flops * 1. / 1e6))
+
+        self.expected_preserve_computation = self.preserve_ratio * self.org_flops
```

Comment:
so AMC is pruning based on flops rather than weights?

Reply:
based on weight for each layer, but the overall target is based on flops.